Add SingleLineStringStyle.PlainExceptAmbiguous and AmbiguousEscapeStyle

russellbanks opened the pull request:
This pull request adds SingleLineStringStyle.PlainExceptAmbiguous that escapes Strings that would be a different type if they were not a String. Resolves #367.

For example, when SingleLineStringStyle is set to PlainExceptAmbiguous, "12" will become "12" in Yaml, whereas Plain would have it be 12.

I have also added an extra configuration option of ambiguousEscapeStyle so the user can choose the escape style for when the string is ambiguous. If it is set to SingleQuoted, then the above example would be escaped like '12', rather than "12". This defaults to DoubleQuoted.

The private function String.isAmbiguous() deems a String as not a String if it was not a String by parsing it as a Long, Double and Boolean. If it is none of these, then it is not ambiguous and we can leave it as plain.

I have written 6 tests that test an Int, Float, and Boolean with each escape style.

It fallbacks to Plain for multiline strings.

I don't tend to contribute to libraries so let me know if I've missed anything or there's anything that can be done better :)

charleskorn commented on src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:
```diff
@@ -83,12 +85,18 @@ public enum class SingleLineStringStyle {
     SingleQuoted,
     Plain,
-    ;
+    PlainExceptAmbiguous;
```
Might be good to add a KDoc comment here to explain what 'ambiguous' means here.

charleskorn commented on src/commonTest/kotlin/com/charleskorn/kaml/YamlWritingTest.kt:
```diff
@@ -200,6 +200,69 @@ class YamlWritingTest : DescribeSpec({
                 }
             }
+
+            context("serializing a string with the value of an integer using SingleLineStringStyle.PlainExceptAmbiguous") {
```
Could you please add some test cases for when `singleLineStringStyle = SingleLineStringStyle.PlainExceptAmbiguous` but the value isn't considered ambiguous? eg. an ordinary string, a value like `1.2.3` that looks numeric but isn't etc.

Also, could you please add some test cases for encoding non-string values with `singleLineStringStyle = SingleLineStringStyle.PlainExceptAmbiguous` to ensure they're not quoted?

charleskorn commented on src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt:
```diff
@@ -180,6 +181,15 @@ internal class YamlOutput(
     }
 
+    private fun String.isAmbiguous(): Boolean {
+        return when {
```
There are other patterns that are ambiguous - take a look at `YamlScalar`'s `convertToIntegerLikeValue`, `toFloat`, `toDouble` and `toBoolean` methods for most of the formats. 

`null`, `~` and empty strings will be interpreted as `null` - these need to be treated as ambiguous as well.

charleskorn reviewed (changes requested):
Thanks for the PR @russellbanks. Overall, this looks good to me.

I've got a couple of suggestions for some more test cases, and there are some other values that are ambiguous that need to be considered as well.

charleskorn commented:
> For the name, PlainEscapedSingle and PlainEscapedDouble could work instead and that would mean ambiguousEscapeStyle could be removed. Let me know what you think @charleskorn

I like the names you've used, and the separation between the two properties makes sense to me.

russellbanks commented:
@charleskorn I've added KDoc for PlainExceptAmbiguous and test cases for when the configuration is PlainExceptAmbiguous and the string is unambiguous and for non-string values.

I've changed the isAmbiguous() function to use toBigIntegerOrNull() instead of toLongOrNull() as otherwise it won't consider strings that represent a value greater than the max value of a Long as ambiguous. I've also added the other patterns from YamlScalar.

charleskorn commented on src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt:
```diff
@@ -180,6 +181,21 @@ internal class YamlOutput(
+        return when {
+            isEmpty() -> true
+            toBigIntegerOrNull() != null -> true
+            startsWith("0x") && removePrefix("0x").toBigIntegerOrNull(16) != null -> true
+            startsWith("0o") && removePrefix("0o").toBigIntegerOrNull(8) != null -> true
```
Anything that starts with `0x` or `0o` is ambiguous, even if the remainder is not a valid integer.

charleskorn commented on src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:
```diff
@@ -83,12 +85,35 @@ public enum class SingleLineStringStyle {
-    ;
+
+    /**
+     * This is the same as [SingleLineStringStyle.Plain], except strings that are not strictly strings in YAML
+     * without quotes will be escaped with the escape style defined in [AmbiguousEscapeStyle].
+     *
+     * True -> Boolean -> Escaped
+     *
+     * 0xAB -> Hexadecimal -> Escaped
+     *
+     * 1 -> Int -> Escaped
+     *
+     * 1.2 -> Float -> Escaped
+     *
+     * 1.2.3 -> Unambiguously a string -> Not escaped
+     *
+     * @since 0.51.0
```
I think this could be a little clearer:

```suggestion
     * This is the same as [SingleLineStringStyle.Plain], except strings that could be misinterpreted as other types
     * will be quoted with the escape style defined in [AmbiguousEscapeStyle].
     *
     * For example, the strings "True", "0xAB", "1" and "1.2" would all be quoted,
     * while "1.2.3" and "abc" would not be quoted.
```

charleskorn commented on src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:
```diff
@@ -83,12 +85,35 @@ public enum class SingleLineStringStyle {
         }
 }
+
+public enum class AmbiguousEscapeStyle {
```
Sorry, just had another thought on the name of this - how about `AmbiguousQuoteStyle`?

charleskorn reviewed (changes requested):
Thanks @russellbanks.

Two small nits and a minor change for hexadecimal and octal values, and then I think this is good to go.

charleskorn commented:
Looks like there are some linting issues as well - run `./gradlew spotlessApply` to fix them.

russellbanks commented:
Done @charleskorn

charleskorn commented:
Thanks @russellbanks!
